### DISCUSSION TITLE
Postgres Support

### DISCUSSION THREAD

**Discussion by lgflorentino** [kimai/kimai#3274] [Questions & Answers]
Hi I'd like to know why the postgres support was stopped?
See #706 & #755 for the previous work on this topic.

edit: I made a [dev environment](https://github.com/lgflorentino/kimai-dev-sandbox) and a work [branch](https://github.com/lgflorentino/kimai2/tree/abstract-db)  if anyone is interested.

**Comment by kevinpapst** [comment]
I am interested to see your progress.
I was fighting with a lot of issues in the past, especially in the places where parameters were passed to the queries.
At least all bool params had to be explicitly marked.
Also I remember issues with the existing SQL statements in the old migrations, which were not abstracted by the Doctrine SchemaTools but written in plain SQL.

**Comment by lgflorentino** [comment]
The SQL commands looked intimidating which is why I asked the question. I'll keep an eye out for the parameter issues, I haven't come across that yet.

**Reply by kevinpapst** [reply]
Yeah I already fixed a lot of issues in the past, so maybe there are not many problems left.
Can you try to run the entire test suite with `composer tests`? 

You will have to `cp phpunit.xml.dist phpunit.xml` first and provide a database connection in https://github.com/kevinpapst/kimai2/blob/master/phpunit.xml.dist#L19 - the database will be filled automatically then.

**Reply by lgflorentino** [reply]
I can run the full test suite. I use the following, where [env](https://github.com/lgflorentino/kimai-dev-sandbox/blob/main/.env)
`source .env; docker-compose run --rm \ 
    -e DATABASE_URL=${MYSQL_DB_URL} composer kimai:tests`
Result:
```php
Time: 04:03.833, Memory: 358.50 MB
There was 1 failure:
1) App\Tests\API\TimesheetControllerTest::testPatchAction
Failed asserting that 25260 matches expected 25200.
/app/tests/API/TimesheetControllerTest.php:600
phpvfscomposer:///app/vendor/phpunit/phpunit/phpunit:97
FAILURES!
Tests: 3028, Assertions: 34267, Failures: 1, Skipped: 3.
Remaining self deprecation notices (6)
Remaining direct deprecation notices (3)
Remaining indirect deprecation notices (6)
Legacy deprecation notices (22)
Other deprecation notices (2)
Script vendor/bin/phpunit tests/ handling the tests event returned with error code 1
Script @tests was called via kimai:tests
```

The failure appears unrelated to me.
 

I've been looking at the old commits as a reference in #755 as well as the doctrine/symfony docs. I've created a test grouping locally `@group database` which I'm using to discover how everything works.

**Reply by kevinpapst** [reply]
I think this is the only relevant commit in the one issue: https://github.com/kevinpapst/kimai2/pull/755/commits/aeb89d01ed5ff56eb6ccdffe8507688913a3ff3d

I can't tell if that was a mistake on my end, because I have no real experience with Postgres, but all bool parameter were not correctly interpreted and the queries failed. Probably try to use some search forms where bools (eg active/inactive for users) are used.
Those might probably fail.

**Reply by batistein** [reply]
@lgflorentino what is the current status on integrating postgres?

**Reply by lgflorentino** [reply]
@batistein I've been able to get the connection working. I've also done some  work recently on rewriting the `addSql()` in the Migrations. There is a portability inconsistency with doctrine's `smallint` and mysql's `tinyint` [Example](https://github.com/lgflorentino/kimai2/blob/720fb62069c83dda77321fc8263595e35b2abead/src/Migrations/Version20180701120000.php#L55). I haven't discovered all of these type of issues yet. It also looks like in some other migrations there are some SQL statements that may not be able to be rewritten easily with doctrine. [ALTER](https://github.com/lgflorentino/kimai2/blob/720fb62069c83dda77321fc8263595e35b2abead/src/Migrations/Version20180715160326.php#L51), UPDATE etc. I believe this might be what @kevinpapst was talking about further up. Finally the other thing that I noticed is that doctrine can use defaults which can change the raw SQL that will be used by the DBAL. This makes it slightly different to the SQL in the `addSQL()` calls. I am yet to resolve this. I've been testing this with `php bin/console doctrine:schema:update --dump-sql` . The changes I have made so far are [here](https://github.com/kimai/kimai/compare/main...lgflorentino:kimai2:abstract-db)

**Reply by batistein** [reply]
awesome thank you very much for your work @lgflorentino 
Is it possible to use your fork already? I saw a lot of progress in the past weeks

**Reply by lgflorentino** [reply]
It cannot be used at this time. Connection & migration code works but the app does not work with the migrated pgsql db yet due to issues documented in #3912 .

**Comment by lgflorentino** [comment]
Hi @kevinpapst , Do you have any advice on making [these ALTER statements](https://github.com/kimai/kimai/blob/a230be77dd36a4e23853cb0bace93fcae1885ed9/migrations/Version20180715160326.php#L49) portable ?  Doctrine deprecated `renameColumns()` and `addSql()` is not executed immediately so I cannot create, copy then drop the columns to replicate the old migration code.

**Reply by kevinpapst** [reply]
Add some methods to [AbstractMigration](https://github.com/kimai/kimai/blob/main/src/Doctrine/AbstractMigration.php):
```php
  protected function isMySQL(): bool {
        $platform = $this->connection->getDatabasePlatform();
        return $platform instanceof MySQLPlatform;
  }

  protected function isPostgreSQL(): bool {
        $platform = $this->connection->getDatabasePlatform();
        return $platform instanceof PostgreSQLPlatform;
  }

```
I doubt that there is something better available then wrapping the code in different if / else 🤷

**Reply by lgflorentino** [reply]
Thanks. I'll try that out.

**Reply by lgflorentino** [reply]
Just checking. Is this file named correctly ? [migrations/Version20993112235958.php](https://github.com/kimai/kimai/blob/main/migrations/Version20993112235958.php) Also what is the reason it doesn't extend `App\Doctrine\AbstractMigration`?

**Reply by kevinpapst** [reply]
a) Why should it be wrong?
b) Mistake, thanks for pointing it out!

**Reply by lgflorentino** [reply]
Wouldn't this migration file be executed last because of the date stamp in the filename [version-numbers](https://www.doctrine-project.org/projects/doctrine-migrations/en/3.5/reference/version-numbers.html#version-numbers)? I was wondering if this is intended or not. For Example: if I created a migration with today's date to replicate the current schema in postgres, this migration would still run after it.  Thank you for the quick response :)

**Reply by kevinpapst** [reply]
Thank you, totally correct 👍 That piece of information was missing and it's too late for guessing 😁 

Actually that was need in the last 12 months, but I forgot to revert it to a proper date. Will be fixed in next beta!

**Reply by lgflorentino** [reply]
I see, No worries :)

**Comment by lgflorentino** [comment]
I've been able to convert the migration SQL here [Version20230131123812.php](https://github.com/lgflorentino/kimai2/blob/abstract-db/migrations/Version20230131123812.php). However I'm now running into a different issue with the importing of test data into the db. 

These are logs of SQL commands that are sent to my pgsql instance
```sh
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.736 UTC [29] LOG:  execute <unnamed>: INSERT INTO kimai2_customers (name, number, comment, visible, billable, company, vat_id, contact, address, country, currency, phone, fax, mobile, email, homepage, timezone, invoice_text, budget, time_budget, budget_type, color, invoice_template_id) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23)
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.736 UTC [29] DETAIL:  parameters: $1 = 'Test', $2 = '1', $3 = 'Test comment', $4 = 't', $5 = 't', $6 = 'Test', $7 = NULL, $8 = 'Test', $9 = 'Test', $10 = 'DE', $11 = 'EUR', $12 = '111', $13 = '222', $14 = '333', $15 = 'test@example.com', $16 = NULL, $17 = 'Europe/Berlin', $18 = NULL, $19 = '1000', $20 = '100000', $21 = NULL, $22 = NULL, $23 = NULL
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.737 UTC [29] LOG:  execute <unnamed>: SELECT CURRVAL($1)
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.737 UTC [29] DETAIL:  parameters: $1 = 'kimai2_customers_id_seq'
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.737 UTC [29] ERROR:  syntax error at or near "end" at character 69
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.737 UTC [29] STATEMENT:  INSERT INTO kimai2_projects (name, order_number, order_date, start, end, timezone, comment, visible, billable, invoice_text, global_activities, budget, time_budget, budget_type, color, customer_id) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)
kimai-pgsql-test-cont  | 2023-02-15 12:37:01.738 UTC [29] LOG:  statement: ROLLBACK

```
The call responsible for the above SQL is [/src/Command/ResetTestCommand.php#L76](https://github.com/lgflorentino/kimai2/blob/ecec0290e06314575d4ce7a99d0bed44c1dbd9a3/src/Command/ResetTestCommand.php#L76)

The create projects table call in the migration is here [migrations/Version20230131123812.php#L605](https://github.com/lgflorentino/kimai2/blob/ecec0290e06314575d4ce7a99d0bed44c1dbd9a3/migrations/Version20230131123812.php#L605)

I'm not sure why it is failing.

**Reply by kevinpapst** [reply]
I have only minimal postgres experience and can't really help. 
Probably `kimai2_customers_id_seq` is invalid?

**Reply by lgflorentino** [reply]
I believe the error is with the insert into the projects table. The Activity Entity and the Customer Entity both succeed before the Project Entity fails. Is there some difference between these Entities ?

**Reply by lgflorentino** [reply]
I'll look into the `kimai2_customers_id_seq`. I've not seen that before.

**Reply by kevinpapst** [reply]
Nice progress btw 👍 

And now I see ` syntax error at or near "end" at character 69` => likely related to the `end` column of the project table.

Probably the column name needs to be quoted, as it seems to be a reserved word in postgres: https://www.postgresql.org/docs/current/sql-keywords-appendix.html

Please try to change the line
https://github.com/kimai/kimai/blob/75aea7c51be3237c1f99a69a61a91e9bd7caf51c/src/Entity/Project.php#L105

And add backticks around `end`:
```
#[ORM\Column(name: '`end`', type: 'datetime', nullable: true)] 
```

**Reply by lgflorentino** [reply]
Thank you ! That got it working :+1:

**Comment by lgflorentino** [comment]
I have been trying to get some tests that will work with `phpunit --group=database`  but ultimately couldn't figure out a nice way to test it. I've gone back to a simple test that is manually run with different `DATABASE_URL`s to see if migrations and/or connections  fail or not.  Please let me know any ideas for testing the migration code that is better than this. After the backtick fix I was able to successfully install with postgres db attached and follow the instructions up to creating an admin user and logging in. As expected once logged in there are db errors. I will start looking into those soon.

**Reply by kevinpapst** [reply]
Not sure what you are looking for. The CI runs these steps to test all migrations:
https://github.com/kimai/kimai/blob/main/.github/workflows/testing.yaml#L102

**Comment by lgflorentino** [comment]
I created #3912 . I'll do some testing to see if using a quote strategy will fix the doctrine exception.

**Comment by llann** [comment]
Hello @lgflorentino nice job! I’m really interested by this Postgres support. Is it still something you’re working on? Thanks!

**Comment by jeroenhabets** [comment]
Hi, 

I am also interested in PostgreSQL support but I would not hold my breath. Though the maintainer is open for [discussion](https://github.com/kimai/kimai/issues/3912#issuecomment-1462271093) and [PRs](https://github.com/kimai/kimai/pull/755#issuecomment-692846188) (which is already quite a big ask when a project only supports MySQL & Maria DB! So kudos for being this welcoming), the community, me and you included, has not been forthcoming. 
Furthermore as the project uses reserved words as column names, adding support is more involved (e.g. https://github.com/kimai/kimai/issues/3912)

So IMHO unless someone is willing to pick up issues like #3912 by @lgflorentino and then complete the journey of making all queries more SQL compliant, or even more radically migrate the database schema to avoid reserved words and then complete that journey, I'd consider the answer to this discussion: PostgreSQL support won't happen.

P.S. In my companies I try to standardize on one single database, and I highly prefer rock-solid extensible PostgreSQL over MariaDB (and won't touch anything owned by Big Red anymore), so stumbled upon this issue surveying potential solutions.

**Reply by kevinpapst** [reply]
Thanks for the summary 👍  

Kimai always was planned to be hostable by "everyone". 
Which means shared hosting as well. 
And that in turn means MySQL / MariaDB.

I don't think that reserved words should be a problem, as all queries (except for the migrations) are using Doctrine Query Language, which escapes (reserved) column names.

I do not see this coming, as no one is willing to work on that task. Its easy to request, yes. But knowing the open source community for quite a while, I would highly doubt that anyone will ever show up and work on that to a point where I could take over.

**Comment by kevinpapst** [comment]
I will lock this discussion now, as there is nothing more to say. 

Get in touch if your company wants to sponsor the support for PostgreSQL.
